### ISSUE TITLE
Improve purpose configuration

### ISSUE THREAD

**Issue by agzam** [syl20bnr/spacemacs#7453]
#### Description :octocat:

Things won't stay in the same window.
#### Reproduction guide :beetle:
- Start Emacs
- Create couple of layouts (persp)
- Create vertical and horizontal window splits
- try changing buffers (especially to non-dedicated ones)
- try jumping to dired from a buffer

_Observed behaviour:_ :eyes: :broken_heart:
Things, instead of popping up in the same window (where the cursor is) often would open in other-window 
Even `SPC Tab` would jump to buffer in other window instead of opening previously used buffer in the same window
#### Additional info:

This is happening only in develop branch I have tried master and there everything works as it used to be. 
#### System Info :computer:
- OS: darwin
- Emacs: 25.1.1
- Spacemacs: 0.200.2
- Spacemacs branch: master (rev. 7bdde17)
- Graphic display: t
- Distribution: spacemacs
- Editing style: vim
- Completion: helm
- Layers:

``` elisp
(csv yaml
     (org :variables org-enable-reveal-js-support t)
     (auto-completion :variables auto-completion-tab-key-behavior 'cycle)
     emacs-lisp javascript html markdown shell-scripts clojure lua react
     (auto-completion :variables auto-completion-enable-snippets-in-popup t auto-completion-enable-help-tooltip t auto-completion-enable-sort-by-usage t)
     syntax-checking spell-checking helm
     (shell :variables shell-enable-smart-eshell t shell-default-term-shell "/bin/zsh")
     (deft :variables deft-directory "~/Google Drive/deft" deft-auto-save-interval 60)
     fasd ibuffer osx chrome
     (semantic :disabled-for emacs-lisp)
     restclient emoji ranger search-engine imenu-list pdf-tools docker git
     (version-control :variables version-control-diff-tool 'diff-hl version-control-global-margin t)
     (github :variables gist-view-gist t)
     agzam ag-dash synonyms)
```

**Comment by bmag** [comment]
This is because of the new `spacemacs-purpose` layer (originally #1958). See the [layer's README](https://github.com/syl20bnr/spacemacs/blob/develop/layers/%2Bspacemacs/spacemacs-purpose/README.org) and [window-purpose](https://github.com/bmag/emacs-purpose) for details.

**Comment by TheBB** [comment]
I'm confused though because the report claims @agzam is on master, but the symptoms perfectly match window-purpose behaviour.

**Comment by bmag** [comment]
The report was made from master, apparently, but according to this [message in gitter by @agzam](https://gitter.im/syl20bnr/spacemacs?at=58052040759f6d2436bf009a) the problem is in develop, not master:
> So yeah… I just tried in master and can confirm the issue with fucked up windows is NOT in master. Something gotten into purpose or layouts layers in develop that’s causing this shit.
basically it’s no longer possible to use develop branch with multiple windows. SPC Tabwouldn’t stay in the same window, SPC B B won’t open buffers in the same window. Jumping to Dired sometimes won’t fucking stay in the same window and for whatever crazy reason would open in the other window

**Comment by agzam** [comment]
@bmag purpose layer was added last week (correct me if I'm wrong) but this stuff appears to be happening fairly recently.

**Comment by bmag** [comment]
It was added less than 24 hours ago, actually.

**Comment by agzam** [comment]
This layer imho should be disabled by default. You guys keep adding layers to spacemacs default distro, this makes extremely difficult for newcomers. People already have to learn layouts, eyebrowse, etc. and now this.

**Comment by syl20bnr** [comment]
Layouts are transparents if not used, it is not necessary to learn them right from the start.
We will configure purpose to have a better and more consistent experience with window placement, if there is too much rough edges right now that it cannot even be used in develop then we can make it opt-in while we fix the issues.

**Comment by agzam** [comment]
I know right? layouts and eybrowse will not do anything until you make a decision to use the, With purpose well, my experience was "goddammit... wtf... why the heck stuff opens in different windows?". And I had no idea that something needed to be turned off.
I'm back in "develop", disabling puppo-mode seems made it less crazy.

**Comment by agzam** [comment]
But disabling the mode is not a solution. I see the benefit of purpose layer. But it should be seamless. For example no window should be "dedicated" by default. If I want to jump to direx-mode or open CIDER-repl or whatever in non-dedicated window - it should let me do that.

**Comment by syl20bnr** [comment]
We have yet to define sensible configuration for purpose, it should be a middle ground between no rule at tall to create new window (like on master) and some rules to set purpose to windows to have the best POLA effect possible, you are right that is should be seamless. We will make it opt-in only temporarily to let us fix biggest regressions.

**Comment by bmag** [comment]
The unexpected effect here is caused because purpose overrides `switch-to-buffer`, so it prefers to open a buffer in a same-purpose window rather the current window. To be more exact, with purpose, `switch-to-buffer` tries to display the buffer with these functions, in this order:
```elisp
purpose-display-reuse-window-buffer
purpose-display-reuse-window-purpose
purpose-display-maybe-same-window
purpose-display-maybe-other-window
purpose-display-maybe-other-frame
purpose-display-maybe-pop-up-window
purpose-display-maybe-pop-up-frame
```
It's configurable, so we can just move `maybe-same-window` to the top, and that will probably fix most of the surprising behaviour.

> For example no window should be "dedicated" by default

No window is dedicated by default, but I think you meant something else. Normally purpose prefers a window with the same purpose as the buffer, even if it's not the current one (as explained in the previous paragraph), and that's what I think you meant. A purpose-dedicated window is a window that will not change it's purpose (won't be used for buffers of other purposes).

**Comment by TheBB** [comment]
If you don't mind I'll turn this issue into a general one for the “purpose” (pun intended) of seamless configuration of purpose.

**Comment by bmag** [comment]
> It's configurable, so we can just move `maybe-same-window` to the top, and that will probably fix most of the surprising behaviour.

Implemented in #7472 with a new dotvariable. @agzam @synic does this PR fix most of the problems?

**Comment by agzam** [comment]
@bmag I pulled latest develop not too long ago. so far, so good! Thank you!

**Comment by bmag** [comment]
I see from #7506 that also the layer's README needs to be updated (at least the install part)

**Comment by quicknir** [comment]
Experiencing the same issue. I'd like to add my 0.02, that this layer should not be installed by default. For people who want to use it, adding a single line is not much effort, and the standard spacemacs layer is already quite large. Overriding `switch-to-buffer` is extremely invasive as well, I'm not sure how easy it will be to eliminate edge cases in its interaction with other packages.

Aside from that, I think the concept of the package is interesting, and it's something I would consider using. I really hope that this is integrated fully with workspaces/layouts. By that I mean that saved workspaces/layouts should also save purposes, and serve as means to organize purposes. So purposes should require minimal keybindings, mostly just setting purpose for a given window. But reading the purpose github page, I'm rather concerned, because purpose seems to be an entire system from the ground up: it has its own ways to switch windows, its own notion of window layouts and saving/loading them, etc. So in essence I'm hoping that: a) usage and learning about purpose is not forced on spacemacs usage, and b) that purpose will play nice with existing spacemacs components, instead of competing with them.

**Comment by bmag** [comment]
> Overriding `switch-to-buffer` is extremely invasive as well, I'm not sure how easy it will be to eliminate edge cases in its interaction with other packages.

True, overriding it is invasive, and is not something that is normally done. If you look at alternatives (like [E2WM](https://github.com/kiwanami/emacs-window-manager)), you'll see they have even more advices and overrides. Unfortunately, there is no other way to customize or extend `switch-to-buffer`. However, #7472 makes `switch-to-buffer` behave almost exactly as it used to, and should eliminate almost all of these edge cases. The only exception is that packages that use `switch-to-buffer` without handling buffer-dedicated windows correctly, probably won't handle purpose-dedicated windows correctly either (IOW purpose-dedication might trigger existing bugs in other packages).

> I think the concept of the package is interesting

Thank you :-)

> saved workspaces/layouts should also save purposes, and serve as means to organize purposes

I don't fully understand what you mean. There is the "purpose configuration", which is the set of rules for deciding what is the "purpose" of any given buffer. Do you mean that each workspace/layout should have its own purpose configuration, and that switching a workspace/layout should also switch the active purpose configuration?

>  I'm rather concerned, because purpose seems to be an entire system from the ground up: it has its own ways to switch windows, its own notion of window layouts and saving/loading them, etc.

If you don't configure `display-buffer`'s behavior (I believe most users don't), then you can continue using Emacs as you did until now, and there will only be three major differences: a) `switch-to-buffer` prefers to reuse a window with the same purpose as the buffer to display than using the current window. b) `pop-to-buffer` is more likely to reuse the current window than another window (that's ok, `pop-to-buffer` is allowed to reuse the current window, it just happens more with Purpose, and is predictable). c) the choice of "other window", when displaying a buffer not in the current window (e.g. `switch-to-buffer-other-window`), is more consistent and easier to prevent (it prefers to reuse windows with the same purpose). Part (a) is eliminated by the aforementioned PR, leaving only (b) and (c), which should be mostly transparent to casual use.

For more advance usage, which usually means configuring `display-buffer-alist`, there is also the concern that Purpose ignores `display-buffer-alist` and offers `purpose-special-action-sequences` instead. Another option is to continue using `display-buffer-alist`, and telling Purpose to ignore the relevant buffers via `purpose-action-function-ignore-buffer-names`. This is a necessary side effect of plugging Purpose into `display-buffer` through `display-buffer-overriding-action`. In the next version of Purpose, which I hope to release within several months, I will plug Purpose through `display-buffer-base-action` instead of `display-buffer-overriding-action`, which means Purpose won't ignore `display-buffer-alist`, and also will let me reduce Purpose's complexity and rely more on the original `display-buffer`.

For advanced usage beyond `display-buffer-alist`, Purpose has more tools for extending it or temporarily turning it off so it's not in the way.

As for Purpose's layouts, there is no necessity to use them, and they do not conflict with anything else. It is just a utility for resotring a "template" of a window layout, for cases where you know what kind of buffers you want to display, but not the specific buffers. Purpose doesn't store any info in the window objects, except for the `purpose-dedicated` window parameter, which is already saved and restored when using saving/loading a workspace/layout. You can continue using workspaces/layouts as you did, and it should just work.

> So in essence I'm hoping that: a) usage and learning about purpose is not forced on spacemacs users, and b) that purpose will play nice with existing spacemacs components, instead of competing with them.

For basic usage, (a) is resolved by the mentioned PR. As for (b), I feel that Purpose already plays nice with existing components and doesn't interfere with them. However, there is still room to integrate Purpose so it *enhances* the other components (but lets get the bugs and the basics straightened out first).

**Comment by quicknir** [comment]
Whoa, well crafted response :-)

> However, #7472 makes switch-to-buffer behave **almost** exactly as it used to, and **should** eliminate almost all of these edge cases

Right? I mean, edge and corner cases are by their nature hard to predict. There's no arguing that adding this layer to the standard set increases the chance of bugs (and reduces performance) by a small amount.

I think in general, that at this point there should be a very high bar to join the standard set of layers, because there's already quite a few, so further increases in size and complexity have to be well justified. Purpose is neat, but I'm not sure it meets that criteria, and the invasive changes definitely push it towards being not installed by default, IMHO. At any rate, I realize there's a good chance I will not be able to convince the maintainers with regards to this particular layer, but I'm hoping at least more generally that this bar for new layers installed by default is kept quite high. (semi-related: an excluded layers option in dotspacemacs might help with this too).

> I don't fully understand what you mean. There is the "purpose configuration", which is the set of rules for deciding what is the "purpose" of any given buffer. 

Example: suppose I'm in workspace 1. I have 4 windows open. I set window 1 to purpose A, and window 2 to purpose B. Now, I switch to workspace 2, which has 3 windows open. Do I have any purposes set in workspace 2? You shouldn't, because workspace 2 is a different config, so if windows 1 and 2 inherited their purpose from worksapce 1 that would be really bad. Another example: let's say that I decided to save the layout of which workspace 1 is a part to a file. I quit emacs. When I start again, I load my layout. I go to workspace 1. Do my windows have purposes?

I feel like until and unless purpose integrates with layouts and workspaces in the way one would naturally expect, that's another strong reason for it not to be in the default distribution.

**Comment by bmag** [comment]
> Right? I mean, edge and corner cases are by their nature hard to predict. There's no arguing that adding this layer to the standard set increases the chance of bugs (and reduces performance) by a small amount.

Right.

By definition (and implementation), a window's purpose is the same as its current buffer. Switching workspaces/layouts restores windows and their buffers, hence it restores purposes as well. A window's purpose can be dedicated (`SPC r P` does that), and this dedication status is also restored when loading a workspace/layout (it's a window parameter, really). If I understood you correctly, then the integration you desire already exists.

**Comment by quicknir** [comment]
Good stuff then, I'm looking forward to trying it out more. If I notice any further purpose-related bugs, I'll post them here rather than open new issues, unless you'd prefer the opposite.

**Comment by rgrinberg** [comment]
Just to confirm, the purpose layer will be opt-in by default, but we will still be able to disable it right?

**Comment by agzam** [comment]
After playing a bit with purpose-mode I have finally decided - it's not for me. I want to be able to open buffers in an exact chosen window. I wasn't sure about adding packages into excluded. I'm with @quicknir here - I wish this layer was not in default distro. But that's just me - maybe someday I will find this layer useful. I remember first time persp-mode/layouts layer was added - I thought it was horrible - now I can't imagine my workflow without it. For now I'll keep this in my user-config: 

    (pupo-mode -1)
    (purpose-mode -1)

**Comment by syl20bnr** [comment]
Purpose is the missing bit to have a complete "spacemacs-layout" feature. As layouts purpose will be transparent if not explicitly used. We have still to work out the integration to achieve this goal. As we did not find the right integration for persp from the beginning, I'm sure it will take some time to find the right integration for purpose. The more people will use it the more we can find a good integration.

To start we need to display the purpose info nicely in the mode-line, I'll open an issue for this particular task.

For now I merged #7472 as it allows to make purpose more transparent which is a step in the right direction but in the final integration I hope we won't need such variable.

**Comment by syl20bnr** [comment]
oh nice: https://github.com/TheBB/spaceline/commit/6d7651ce1e2b1170378b8660f6dc6a553a83b893

Thanks @TheBB :-)

**Comment by syl20bnr** [comment]
#7472 has been pushed.

I may create a new thread on purpose in a near future, still need to play with it a lot.

**Comment by TheBB** [comment]
> I may create a new thread on purpose

This package name has great pun potential, @bmag.

**Comment by rgrinberg** [comment]
Supposedly there's an incoming change that will make purpose mode far less
intrusive. I will make my verdict on it once it comes out.

However, I do think that it's invasive enough for people to have an obvious way
of opting out.

**Comment by TheBB** [comment]
> I will make my verdict on it once it comes out. 

It's out…

**Comment by quicknir** [comment]
Purpose is still intrusive by default. Two examples:

 1. Let's say I start compiling. Then i switch to the compilation buffer/window, and I use `SPC b b` to try to get another buffer to show up there. It does not work, and instead of opening the buffer in that window, I switch to another window with that buffer.

 2. Purpose shows the purpose in the modeline, even though I haven't run a single purpose command. This contradicts the display of layouts and workspaces, which are completely transparent unless you start creating more workspaces/layouts. This wastes modeline space. Ideally, there should be a "default" purpose that does not need to be displayed, even when purposes are being used (like with layouts). But at the minimum, if you're not using purposes, these should not appear (like with workspaces).

purpose needs to be opt-in until these issues are worked out.

**Comment by nixmaniack** [comment]
@quicknir Your point number `1` is exactly the same purpose for which purpose was created. It might be unintuitive in the start but once you get used to it and understand its purpose - you have better control on where new buffers/windows pop up. 

All this time we're working(/had been working)/munching code in emacs without purpose so it's unintuitive and I agree on that.  

I, for one, would have opted for keeping it default since this is one such feature which unless you force users to get to know about its existence and make them use it - it won't be discoverable and will remain -- unintuitive. 

And because of above argument - keeping the purpose in modeline makes sense from the start itself, IMO. 

> purpose should be opt-in until these issues are worked out.

The new dotvariable on develop, I think mimics the default emacs behaviour, I haven't tested it though.

**Comment by bmag** [comment]
@quicknir point 1 should have already been fixed in latest develop, with the mentioned dotvariable. The value of `(cdr (assq 'switch-to-buffer purpose-action-sequences))` should now be:
```elisp
(purpose-display-maybe-same-window
 purpose-display-reuse-window-buffer
 purpose-display-reuse-window-purpose
 purpose-display-maybe-other-window
 purpose-display-maybe-other-frame
 purpose-display-maybe-pop-up-window
 purpose-display-maybe-pop-up-frame)
```
This should make `SPC b b` open in current window, unless current window is dedicated to its buffer (indicated by `#` to the right of the purpose in the mode-line) or to its purpose (`!` in the mode-line).

About point 2, the mode-line also shows some active minor-modes - I didn't opt-in to them either (and don't find them helpful). I don't think we need to remove the purpose indicator.

**Comment by quicknir** [comment]
@bmag I'm on latest develop, and I have `dotspacemacs-switch-to-buffer-prefers-purpose` nil, and I still see the behavior that I mentioned. Did you try to reproduce and fail?

As for 2, I don't really see the relevance of your point. If you think most people would prefer eliminating the minor mode display, or having an option, by all means open an issue, and we can see if other people feel the same way.

More on topic, obviously a decision needs to be made what needs to be included and excluded from the modeline. You have not presented any arguments why the purpose indicator needs to be there, when no purpose has been set. I've presented arguments against, but more formally:

 1. Purpose is more similar to layouts and workspaces, which not everybody uses, than to minor modes which everybody uses (even if you don't happen to like an explicit display). If you don't use them, the display adds no value, and is merely confusing, which is why layouts and workspaces don't affect the modeline if not used. Purpose should do the same.

 2. Even for somebody actively using purpose, it seems to me that "edit!" (a locked purpose of edit) can be replaced with "edit", and "edit" can be replaced with a blank. If purpose is transparent when no purpose has been set, then it seems reasonable to not bother displaying a purpose until one has been locked/set. The nice thing about this approach is that it gives you 1 for free. Is there a reason that you don't like this solution?

**Comment by quicknir** [comment]
Recent discovery: purpose currently (even with prefers-purpose set to nil) breaks ranger pretty bad. Basically when it's running in it's default 3 window mode (parent, current, preview), once you move the cursor over a file the right window starts to preview a file. When you move the cursor back to a directory, instead of show the dir contents on the right as it should, it shows it on the left. None of the windows had an ! or a # on the purpose. The only solutions I found were to either use deer (a minimal version of ranger which uses only one window) or to turn off purpose.

**Comment by bmag** [comment]
Sorry for late reply, was very busy. Here goes:

I tried but couldn't reproduce the ranger issue. Anyway, I believe [these two lines](https://github.com/ralesi/ranger.el/blob/master/ranger.el#L1838-L1839):
```elisp
(with-selected-window preview-window
  (switch-to-buffer preview-buffer))
```
should be replace by:
```elisp
(set-window-buffer preview-window preview-buffer)
```
Or at the very least, `switch-to-buffer` needs a `force-same-window` argument.

Regarding the modeline, I meant to say that it isn't a blocker - we can keep Purpose as opt-out even with this modeline issue. 

Now, wether we should remove Purpose from the modeline, that depends on how we want to use Purpose. I see two different usage patterns:

The first usage pattern says that the user knows in advance what purpose they want for each type of buffer. They have their purpose configuration (or use the default one), they know where a buffer will open when they open a buffer (no matter how they open it). If they want a specific window arrangement, they either load a saved window layout, or temporarily disable purpose-mode to make the changes and then reactivate it. In this pattern, the user doesn't necessarily use dedicated windows, but they do rely on the modeline info to know where a buffer will be displayed. They need the info to be there. This is the original usage I had in mind when writing Purpose, and also when writing the layer for Spacemacs.

The second usage pattern says that the user doesn't want Purpose to interfere at all, unless they expilictly set a window's purpose. I believe this is the usage you (and others) have in mind, and in this case the user needs the modeline info only for dedicated windows.

The modeline decision really depends on what usage pattern we want to support (and how we want to integrate it with layouts and workspaces). The current implementation supports the first pattern, but I think it also has decent support for the second pattern (as long as you don't want to change purpose configuration on the fly). Currently, it makes sense to keep the modeline info as it is, because it's relevant for using purpose. Frankly though, I have to say that I'm ok with making Purpose an opt-in feature - after all that was my original intention when writing the layer.

You also wrote about the popup window, that you can't reuse the compilation popup window to display different buffers. That's expected - it's a popup window, it's intended to only be a popup. If you want to "un-popup" it, you can undedicate the window (`SPC w t` toggles buffer-dedication) and then reuse it for other purposes. Well, actually pupo-mode (the replacement of popwin-mode) will still close the window if you press `C-g`, because it doesn't check if the window stopped being a popup window, but I'll fix this issue tomorrow (if all goes right).

**Comment by quicknir** [comment]
@bmag I'll try your fix for ranger, thanks. If it works I'll PR it.

As for the modeline, I had the following idea which seems to fit what you're saying. The segment would take the string corresponding to what is now showing up in the modeline, but before returning it would check two things:

 1. If dotspacemacs-switch-to-buffer-prefers-purpose, return it unchanged.
 2. If it ends with ! or #, return it unchanged.
 3. Else empty.

It seems to me, having explicitly identified two usage paterns, that the value of `dotspacemacs-switch-to-buffer-prefers-purpose` should line up with these two patterns. nil should be usage 2, t should be usage 1. So usage 1 users always get their purpose, as you like. usage 2 users only get the modeline if explicitly dedicated. Everyone's happy.

If my understanding of the variable is incorrect, maybe we could change it to so that it better lines up with the two use cases (which might involve setting more than one option within purpose), and then apply the same logic for the modeline to make everyone happy, with very minimal config.

Cheers.

**Comment by aroig** [comment]
Do you keep track somewhere of stuff broken in conjunction with the new purpose package? #6911 is one of them. Both `company-coq` and `proof-general` interact badly with `window-purpose`.

**Comment by bmag** [comment]
@aroig yes, this issue is for collecting all integration problems. Can you give some more details on how you trigger the issues you described in #6911? There are ways to configure `window-purpose` to behave nicely with other packages, depending on the nature of the conflicts.

**Comment by aroig** [comment]
@bmag Ok to trigger this you need coq in your machine. At least for the first one. Here is how to trigger the problems:

Enable the coq layer in #6911 and open a file `test.v` like this:

    Definition divides (a b : nat) := exists c, a * c = b.
    
    Definition prime (n : nat) := forall a, divides a n -> a = 1 \/ a = n.

The two problems I found are the following:

1. Doing `SPC m ]` evaluates the file until point. This should open a couple of windows (with buffers `*goals*` and `*response*`), showing state of the engine. When `window-purpose` is enabled, this just kills my emacs frame.

2. Doing autocompletion (try to write `Definition` and wait for company to kick in) it should open a documentation buffer on the side, but with `window-purpose` enabled it just opens the buffer in place of `test.v`.

I imagine that for problem 2, we should match the documentation buffer somehow and set an appropriate purpose, for it? Should it be done in the coq layer?

Problem 1 above might be trickier, since proof-general insists in setting the layout on the windows it creates. What should be the general approach for a mode that insists in managing windows on its own?

Note that "fixing the mode" might not be a realistic approach...

**Comment by bmag** [comment]
> Doing `SPC m ]` evaluates the file until point. This should open a couple of windows (with buffers `*goals*` and `*response*`), showing state of the engine. When `window-purpose` is enabled, this just kills my emacs frame.

Could you give a "before/after" of how the screen looks? For example, before might be:
```
+------------+------------+
|            |            |
| myfile.txt |   *Help*   |
|            |            |
+------------+------------+
```
When you "kills my emacs frame", do you literally mean the frame is deleted? It might be because the window layout is incorrect when `proof-delete-all-associated-windows` is called:
```elisp
(defun proof-delete-all-associated-windows ()
  "Delete windows (and maybe frames) showing associated buffers.
Delete a frame if it displays only associated buffers, unless it
is the only frame (try to bury buffers then)."
  (mapc (lambda (w)
	  ;; try to delete window, or frame, or only bury buffer
	  (if (not (frame-root-window-p w)) (delete-window w)
	    (if (< 1 (length (frame-list))) (delete-frame (window-frame w))
	      (window--display-buffer (other-buffer) w 'window))))
	(proof-associated-windows t)))
```

> Doing autocompletion (try to write `Definition` and wait for company to kick in) it should open a documentation buffer on the side, but with `window-purpose` enabled it just opens the buffer in place of `test.v`.

I'm having some trouble following company-coq's code. It's possible that PG gets confused, sets `test.v`'s buffer as `proof-goals-buffer`, and that causes company-coq to show the documentation in the wrong window. Can you provide a "before/after" here as well?

**Comment by aroig** [comment]
> I've just tested it without daemon now, and it works as expected

Well, I meant the layout is as expected. All three windows get the purpose `general`. I'm not sure if this is the right thing.

**Comment by bmag** [comment]
> Is proof-delete-all-associated-windows called on SPC m ] ?

It might (I didn't install Coq, I'm just reading the code), because `proof-layout-windows` calls `proof-delete-all-associated-windows`. The code is hard to follow, but I think `proof-layout-windows` is indeed called by `SPC m ]`. (and the difference between daemon and non-daemon supports this)

> All three windows get the purpose general. I'm not sure if this is the right thing.

Probably not. Can you see if configuring the relevant buffers to have non-default purposes (e.g. response buffers should probably have a `response` purpose, etc.) changes the behavior? Configuration can be accomplished by `purpose-add-user-purposes`, for example:
```elisp
(purpose-add-user-purposes :modes 
                           '((proof-response-mode . response)
                             (proof-goals-mode . goals)))
```

If you know the names of PG's buffers, the variable `purpose-action-function-ignore-buffer-names` can be used. It tells window-purpose to ignore certain buffers and use Emacs' original display rules. For example:
```elisp
(add-to-list purpose-action-function-ignore-buffer-names "^\\*proof-")
```

**Comment by aroig** [comment]
> (purpose-add-user-purposes :modes 
                           '((proof-response-mode . response)
                             (proof-goals-mode . goals)))

Cool! that fixes it.

What should be the purpose of documentation buffers?

Thanks a lot for your time, by the way.

**Comment by bmag** [comment]
> Cool! that fixes it.

Good :smile: For future reference, please comment on the PR what you had to configure.

> What should be the purpose of documentation buffers?

We don't have a special purpose for doc buffers, so it's up to you. `doc` or `help` sounds appropriate.

**Comment by edrex** [comment]
@bmag, sorry for resurrecting an old thread - this seemed like the best place to ask.

I'm seeing some buffers getting an unexpected (for me) purpose:

 - files with an unknown type (foo.bar) get Fundamental mode and get "general" purpose (I would expect "edit")
 - The same happens with protobuf-mode (https://github.com/syl20bnr/spacemacs/pull/8234)

What do you think is going on here? I'm happy to submit a patch if appropriate.

Filed as https://github.com/syl20bnr/spacemacs/issues/8238

**Comment by agzam** [comment]
I'm closing this due to inactivity.
